### DISCUSSION TITLE
xyChart features

### DISCUSSION THREAD

**Discussion by Piratepie** [mermaid-js/mermaid#5326] [Ideas]
TL;DR: ideas on how to implement multiple bars, a legend, and points of interest 

I've run into a couple of limitations using xyChart and I've come up with some suggestions here.

## Multiple bars
A very common use case is showing bars over or next to each other. Here follows 2 examples:

![image](https://github.com/mermaid-js/mermaid/assets/27888345/ed728ab4-1be7-4fed-b7e1-76c005edf88d)
![image](https://github.com/mermaid-js/mermaid/assets/27888345/c0d80f22-aa69-4470-98f3-b517378c3fa6)

Currently, xyChart does take in 2 bar arguments, but doesn't display them in a way that's useful. Recreating the first chart in mermaid yields the following:

```
xychart-beta
    title Number of Students in each group
    x-axis "numbersX" ["Group A", "Group B", "Group C", "Group D"]
    y-axis "numbersY" 0 --> 40
    bar [10,20,20,40]
    bar [20,30,25,30]
```
![image](https://github.com/mermaid-js/mermaid/assets/27888345/567bbb86-9b20-4f03-9091-a147a744e42d)

The bars overlap, hiding the smaller value. It would be great for direct comparisons if either:
- Bars could be placed side by side
- Bars could be placed on top of each other without hiding the smallest value

The example of the 2nd bar can be done in the current implementation, but without a legend becomes hard to use. Which leads me to...

## Legend
There's currently no support for a legends, which makes reading the chart hard. When using multiple lines and/or bars it makes a chart barely useable.

In this example of a line xyChart there is no way of knowing which line is what.

```
xychart-beta
    title "Number of Students in each group"
    x-axis "numbersX" ["Group A", "Group B", "Group C", "Group D"]
    y-axis "numbersY" 0 --> 40
    line [10,20,20,40]
    line [20,30,25,30]
    line [40,30,20,10]
```
![image](https://github.com/mermaid-js/mermaid/assets/27888345/0bad2838-8a1d-4696-a57f-a0c132a417f1)

## Plotted points
Another, more niche, use case of a line chart is to add points to project growth for example. This example shows the real data with lines and the projected usages with points plotted on the chart.

![image](https://github.com/mermaid-js/mermaid/assets/27888345/f2a199f8-d6a3-4337-bc0a-42ca1e9a80d6)

## Suggested solution
A solution would be to add IDs to the `bar` and `line` element. This could be:

_Naming the element like [pie charts](https://mermaid.js.org/syntax/pie.html) or like [gantt charts](https://mermaid.js.org/syntax/gantt.html)_
```
    "Girls" : bar [10,20,20,40]
    "Boys" : line [20,30,25,30]
```

_Adding the ID as a parameter like [sequence diagrams](https://mermaid.js.org/syntax/sequenceDiagram.html), allows for using the keyword "as"_
```
    bar Girls [10,20,20,40]
    line Boys [20,30,25,30]
```

_Wrapping the element with an ID like [flowcharts](https://mermaid.js.org/syntax/flowchart.html)_
```
    girls(bar [10,20,20,40])
    boys(line [20,30,25,30])
```

This would help with differentiating between bars and lines, which could also be used for a legend like that [already implemented in the pie chart](https://mermaid.js.org/syntax/pie.html#pie-chart-diagrams)

As for the plotted points, this could be a useful new element of xyChart which could look like this (including the ID idea mention above):
```
    %% id: string : point [xIndex: int, yValue: float]
    %% projected growth or shrinking in the month december would look like

xychart-beta
    title "Sales projection"
    x-axis [jan, feb, mar, apr, may, jun, jul, aug, sep, oct, nov, dec]
    y-axis "sales" 50 --> 150
    "Grow": point [11, 110]
    "Shrink": point [11, 90]
```

I also saw I'm not the only one, #5207 also mentioned the legend for multiple lines.

**Comment by dschulzg** [comment]
I just discovered mermaid in our gitlab installation.  I'm so impressed!  SOOOO Useful! The above points are the only thing preventing me from being able to inline some charts in a markdown document!.

**Comment by sarod** [comment]
Legend is the one that is lacking the most I find when using multiple line or bars

**Comment by josecastillolema** [comment]
The ability to display values above the bars would be awesome also.

**Comment by NelsonBN** [comment]
The idea of a new type of chat, 'Point Chat' would be really amazing.

**Comment by rhelmot** [comment]
Similar to above, scatter plots would be nice to have. 

I would also like to be able to use dates as range values. Currently, you can only use discrete values with individual dates.

**Comment by xdumaine** [comment]
I added legends in #7724
